src/effect/tap_tempo.rs

```diff
@@ -0,0 +1,151 @@
+use crate::{audio::midi::Message, config::NoteOn, effect::Tempo};
+use wmidi::MidiMessage;
+use State::*;
+
+#[derive(Debug)]
+enum State {
+    Start,
+    TappedOnce(u64),
+}
+
+#[derive(Debug)]
+pub struct TapTempo {
+    note_on: NoteOn,
+    state: State,
+}
+
+impl TapTempo {
+    const MAX_DELAY_BETWEEN_TAPS: u64 = 2_000_000; // microseconds
+
+    pub fn new(note_on: NoteOn) -> Self {
+        Self {
+            note_on,
+            state: State::Start,
+        }
+    }
+
+    pub fn handle_messages(&mut self, messages: &[Message]) -> Option<Tempo> {
+        messages
+            .iter()
+            .flat_map(|message| self.handle_message(message))
+            .last()
+    }
+
+    fn handle_message(&mut self, message: &Message) -> Option<Tempo> {
+        match message.message {
+            MidiMessage::NoteOn(channel, note, _)
+                if channel == *self.note_on.channel && note == *self.note_on.note =>
+            {
+                self.handle_tap(message.timestamp)
+            }
+            _ => None,
+        }
+    }
+
+    fn handle_tap(&mut self, timestamp: u64) -> Option<Tempo> {
+        match self.state {
+            Start => {
+                self.state = TappedOnce(timestamp);
+                None
+            }
+            TappedOnce(last_tap) => {
+                let time_since_last_tap = timestamp - last_tap;
+
+                if time_since_last_tap <= Self::MAX_DELAY_BETWEEN_TAPS {
+                    self.state = TappedOnce(timestamp);
+                    Some(Tempo::new(timestamp, time_since_last_tap))
+                } else {
+                    self.state = TappedOnce(timestamp);
+                    None
```

Comment:
It's been too long between taps, so don't produce a new tempo, but do record the tap. (Not sure how this usually works on real pedals.)